contributors/design-proposals/api-machinery/customresource-conversion-webhook.md

```diff
@@ -0,0 +1,859 @@
+
+1. The stored custom resources in etcd will not be deleted.
+
+2. Any clients that try to get the custom resources will get a 500 (internal server error). this is distinguishable from a deleted object for get and the list operation will also fail. That means the CRD is not served at all and Clients that try to garbage collect related resources to missing CRs should be aware of this. 
+
+3. Any client (e.g. controller) that tries to list the resource (in preparation for watching it) will get a 500 (this is distinguishable from an empty list or a 404).
+
+4. If the user rolls forward again, then custom resources will be served again.
+
+If a user does not use the webhook feature but uses the versioned schema, additionalPrinterColumns, and/or subresources and rollback to a version that does not support them per version, any value set per version will be ignored and only values in top level spec.* will be honor.
+
+Please note that any of the fields added in this design that is not supported in previous kubernetes releases can be removed on an update operation (e.g. status update). The kubernetes release where defined the types but gate them with an alpha feature gate, however, can keep these fields but ignore there value.
+
+### Webhook Request/Response
+
+The Conversion request and response would be similar to [Admission webhooks](https://github.com/kubernetes/kubernetes/blob/951962512b9cfe15b25e9c715a5f33f088854f97/staging/src/k8s.io/api/admission/v1beta1/types.go#L29). The AdmissionReview seems to be redundant but used by other Webhook APIs and added here for consistency.
+
+```golang
+// ConversionReview describes a conversion request/response.
+type ConversionReview struct {
+  metav1.TypeMeta
+  // Request describes the attributes for the conversion request.
+  // +optional
+  Request *ConversionRequest
+  // Response describes the attributes for the conversion response.
+  // +optional
+  Response *ConversionResponse
+}
+
+type ConversionRequest struct {
+  // UID is an identifier for the individual request/response. Useful for logging.
+  UID types.UID
+  // The version to convert given object to. E.g. "stable.example.com/v1"
+  APIVersion string
+  // Object is the CRD object to be converted.
+  Object runtime.RawExtension
+}
+
+type ConversionResponse struct {
+  // UID is an identifier for the individual request/response.
+  // This should be copied over from the corresponding ConversionRequest.
+  UID types.UID
+  // ConvertedObject is the converted version of request.Object.
+  ConvertedObject runtime.RawExtension
+}
+```
+
+If the conversion is failed, the webhook should fail the HTTP request with a proper error code and message that will be used to create a status error for the original API caller.
```

Comment:
Please explicitly define the error return type, e.g. by adding an `Error string` to the ConversionResponse type, or by requiring it to send back a metav1.Status object.